KEP-5547: Update job integration to use workloadbuilder (#6153)
kubernetes/enhancements: keps/sig-apps/5547-integrate-workload-with-job/README.md

  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) within one minor version of promotion to GA
- [x] (R) Production readiness review completed

> - [ ] (R) Production readiness review approved

Reviewer: ```suggestion
- [x] (R) Production readiness review approved
```

- Compile `spec.scheduling` into `Workload`/`PodGroup` objects via the shared `workloadbuilder`
  library instead of bespoke controller logic.
- Support mutable `spec.scheduling.policy.gang.minCount` for elastic scaling, while keeping all
  other `spec.scheduling` fields immutable after creation. This relies on [KEP-4671] that makes 
  `minCount` in `PodGroup`/`PodGroupTemplate` mutable in v1.37 to support workload scaling.
- Skip `Workload`/`PodGroup` creation when a higher-level controller owns the workload (the Job
  carries an `OwnerReference` to a registered parent workload), preserving the
  root-controller-as-compiler principle.
- Ensure proper ordering of `Workload` → `PodGroup` → `Pod` creation.

### Non-Goals

- Multi-level / nested composite (`CompositePodGroup`) structures, since this KEP covers 
  single-level, flat `Job` workloads only.
- Implementing the integration in composite controllers (`JobSet`, `LWS`, `TrainJob`). Those 
  are pursued independently in their own repositories.
- Defining the `scheduling.k8s.io` building-block API or the `workloadbuilder` library 
  itself (owned by [KEP-6089] and consumed here).

## Proposal

> This proposal builds on the recently introduced Workload-aware Scheduling enhancements. We assume

Reviewer: nit: lines 153 - 159, I would just remove.

We have a section that says see also.

Reviewer 2: Theoretically - yes, but only a few nerds look into that section 😅 , so I don't mind this being explicitly called out here.

Reviewer: Ha. Fair enough.

  are pursued independently in their own repositories.
- Defining the `scheduling.k8s.io` building-block API or the `workloadbuilder` library 
  itself (owned by [KEP-6089] and consumed here).

## Proposal

This proposal builds on the recently introduced Workload-aware Scheduling enhancements. We assume
the reader is acquainted with the following KEPs:

- [KEP-4671]: Gang Scheduling.
- [KEP-5710]: Workload-aware preemption.
- [KEP-5732]: Topology-aware workload scheduling.
- [KEP-6089]: WAS Controller APIs.

The Job controller is extended to compile the user's scheduling intent into `Workload` and
`PodGroup` objects as part of its pod-management lifecycle, so that pods belonging to a Job are
scheduled according to the requested policy before they are created. The intent is expressed
through a new `spec.scheduling` field.

The key design principles for this alpha are:

- One `Job` maps to one `PodGroup` representing a single homogeneous group of pods (a flat,
  single-level structure). For a root Job, the controller also compiles the enclosing `Workload`;
  for a non-root Job whose parent owns the `Workload`, that single `PodGroup` instead attaches to

>   the parent's compiled `Workload`/`CompositePodGroup`.

Reviewer 3: This is slightly misleading:
1) We should always link to Workload - let's clarify that.
2) We link to parent object (CPG) based on the existence of the annotation.

Author: Updated. PTAL.

  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) within one minor version of promotion to GA
- [x] (R) Production readiness review completed
- [x] (R) Production readiness review approved
- [x] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

## Summary

This KEP integrates the Workload-aware Scheduling (WAS) APIs (`Workload` and `PodGroup`) into the

> batch/v1` Job by adding a user-facing `spec.scheduling` field, allowing users to express explicit

Reviewer 2: ```suggestion
`batch/v1` Job by adding a user-facing `spec.scheduling` field, allowing users to express explicit
```

> core `batch/v1` Job by adding a user-facing `spec.scheduling` field, letting users express explicit

Reviewer 2: ```suggestion
`batch/v1` Job by adding a user-facing `spec.scheduling` field, allowing users to express explicit
```

scheduling intent such as gang scheduling[^1], topology co-location, and disruption policies.

> The first alpha was introduced in v1.36 intentionally bypassed this user-facing API: the controller inferred a

Reviewer 2: ```suggestion
The first alpha, introduced in v1.36, intentionally bypassed this user-facing API: the controller inferred a
```

scheduling constraints. This is a challenge for parallel applications (i.e., AI/ML training
workloads, MPI jobs) that require all pods to be scheduled and run together or none (gang
scheduling[^1]). The v1.36 alpha brought gang scheduling to the Job controller, but it did so by
inferring a hardcoded `Gang` policy from the Job's type rather than from explicit user intent.

Users have diverse use cases and require the ability to express explicit intent, such as opting
in or out of gang scheduling, requesting specific topologies, or configuring disruption policies
for their workloads. [KEP-6089] standardizes the reusable scheduling building blocks
(`scheduling.k8s.io/v1alpha3`) and a shared `workloadbuilder` translation library so that workload
controllers can expose these features consistently. This KEP integrates those building blocks into
the core `Job` API, "blazing the path" for the rest of the ecosystem while resolving the usability
gaps of the initial alpha.

### Goals

- Add a user-facing `spec.scheduling` (`JobSchedulingConfiguration`) field to the `batch/v1` Job,
  embedding the `scheduling.k8s.io/v1alpha3` building blocks (`policy`, `constraints`,

>   `disruptionMode`, `resourceClaims`) so users express explicit scheduling intent.

Reviewer 2: ```suggestion
  `disruptionMode`, `resourceClaims`) so users can express explicit scheduling intent.
```

- Default to `Basic` (standard pod-by-pod) scheduling when `spec.scheduling` is omitted, so the
  runtime scheduling behavior of existing Jobs is unchanged (implicit opt-out from gang). Following
  the [KEP-6089] "Universal Representation" principle, the controller still materializes a `Basic`

>   `Workload`/`PodGroup` for these Jobs.

Reviewer 4: Basic policy in a PodGroup alters the behavior of scheduler. In particular, pods from a basic pod group are queued and scheduled together, including workload-aware preemption attempt if needed. The difference from Gang policy is that scheduler doesn't enforce any `minCount`, so any number of schedulable pods go to binding.

I'm not sure if given this information we can say the scheduling behavior is **unchanged**, but from the user perspective, the change probably won't be that visible

Reviewer 2: So, w/o `minCount` enforcement, the user-visible behavior is pod-by-pod, is that correct way of saying it? If it is, I'm not worried about that.

Reviewer 4: From highest level perspective, it will be the pod-by-pod  behavior, i.e., if there is a space for one pod, it will get scheduled, regardless of schedulability of others pods from the Job.

I think the mostly visible divergence will be with preemption. In particular, when some pods are schedulable within the cycle, but for others preemption is needed,  the scheduler will try the preemption attempt, delaying the startup time of schedulable pods (because all pods have to come back to the scheduling queue waiting for the victims to be removed). Design:
https://github.com/kubernetes/enhancements/blob/a201960e69627ab0dbbde5866311ea4bc1dc7e7d/keps/sig-scheduling/4671-gang-scheduling/README.md?plain=1#L1209-L1221

Also, the workload-aware preemption will be tried for basic pods, what may lead to somewhat different preemption results than with the pod-by-pod preemption, because WAP considers the whole cluster and pod group at once instead of each node and pod separately.

Reviewer 2: Thanks Maciek for that explanation, that definitely sheds more light into what we can expect. I believe the question then turns into how big the delay can be, and I know the answer is something along: "it depends on how many pods we're evicting and how fast the entire preemption will be".

Reviewer 3: That's a great point @macsko !

And so, now I'm wondering if we didn't unify that too much and for "subsequent scheduling attempts", we should actually diverge it based on the policy:
1) For gang - keep what we have in the KEP (for reasons that we discussed already)
2) For Basic - proceed with binding if we managed to schedule any pod.

(2) would make it more "backward compatible" which seems highly desired here.

Reviewer 3: @Argh4k

Reviewer 4: > 1. For gang - keep what we have in the KEP (for reasons that we discussed already)
> 2. For Basic - proceed with binding if we managed to schedule any pod.

That sounds like a good idea. This way, basic policy would be as close to the pod-by-pod behavior as possible. 

However, I'm unsure how could we differentiate such behavior in the scheduler's code, given the gang policy is handled by the plugin. One idea would be to allow the basic policy to be accessed directly from the scheduling cycle.

Reviewer 3: Yeah - I think that we would need the framework to interpret the scheduling policy, which I agree is not ideal.
[The alternative would be some other extension point to reflect whether WAP should be used or not, but that seems very artificial to me....]

@macsko - will you be able to open an update PR to gang scheduling KEP to suggest that?

Reviewer 4: Yes, I'll open a PR

Reviewer 4: Opened https://github.com/kubernetes/enhancements/pull/6194

Author: > I'm not sure if given this information we can say the scheduling behavior is unchanged, but from the user perspective, the change probably won't be that visible

Updated the basic behavior across the KEP. PTAL.

- Let users opt in to `Gang` scheduling, with `minCount` defaulting to `parallelism` (or
  `completions` when parallelism is unset) when omitted.
- Compile `spec.scheduling` into `Workload`/`PodGroup` objects via the shared `workloadbuilder`
  library instead of bespoke controller logic.
- Support mutable `spec.scheduling.policy.gang.minCount` for elastic scaling, while keeping all
  other `spec.scheduling` fields immutable after creation. This relies on [KEP-4671] that makes 
  `minCount` in `PodGroup`/`PodGroupTemplate` mutable in v1.37 to support workload scaling.
- Skip `Workload`/`PodGroup` creation when a higher-level controller owns the workload (the Job

>   carries an `OwnerReference` to a parent controller that itself creates and manages the

Reviewer 3: This is very misleading.

While we may want to skip `Workload` management if this is not the top-level workload, it doesn't necessarily mean we want to skip the PodGroup creation too.
There are definitely cases, where non-root true-workload controller will actually be managing PodGroups (e.g. either JobSet or Job would have to manage PodGroups if they are part of TrainJob, as it doesn't even know about Jobs themselves).

This is exactly what podgroup-template annotation is for:
https://github.com/mm4tt/enhancements/blob/5904ead9e0c632088a050361d6f1d0fb6a3b7fb2/keps/sig-scheduling/6089-was-controller-apis/README.md#the-solution-downward-mapping-annotations

Let's clarify that somehow.

Reviewer 2: Reading through https://github.com/mm4tt/enhancements/blob/5904ead9e0c632088a050361d6f1d0fb6a3b7fb2/keps/sig-scheduling/6089-was-controller-apis/README.md#1-runtime-podgroup-and-compositepodgroup-lifecycle-management (which is not even merged) I see the goal is to verify the future path. So the API will allow both Centralized and Delegated management, and we're planning to wait for the the feedback before moving one or the other way. Any particular reason we want both for job integration in that case?

Reviewer 3: That's actually great point and we should discuss it. I think there are two main aspects here:

1) Do we really need both?
If you're top-level controller, then indeed you may explore which option works better for you.
[Though I would claim that we already know the cases (it's effectively any example that has 3+ layers, e.g. TrainJob with TrainJob -> JobSet -> Job), where we know that some kind of delegation will be needed. It's basically because as a TrainJob, you're not managing the Jobs themselves (you only transitively do that via JobSets), so if we don't manage the control object, how can we even manage the corresponding scheduling-oriented object?]
But Job while it can be top level controller, it can also be a lower-level one (as in the example below).
So if we want to allow e.g. a JobSet to experiment with both modes, we need to support both modes in the Job (because delegation from JobSet requires supporting the "I'm the delegate" mode in the job controller).

2) Should this be part of this release?
The reason I'm advocating for it is that if we want higher level controller to experiment with both modes, we kind-of need to support both modes on lower-level ones (otherwise we're effectively forcing them to one of the modes). Given that Job is lower-level controller for multiple ongoing integrations (JobSet, TrainJob), Job I think we should try to prove both modes here.

Author: > Should this be part of this release?
The reason I'm advocating for it is that if we want higher level controller to experiment with both modes, we kind-of need to support both modes on lower-level ones (otherwise we're effectively forcing them to one of the modes). Given that Job is lower-level controller for multiple ongoing integrations (JobSet, TrainJob), Job I think we should try to prove both modes here.

+1. Updated the KEP accordingly. PTAL.

Reviewer 3: FWIW - I pinged @soltysh on Friday to double check with him that he is also ok with my reasoning here and he confirmed.

- Ensure proper ordering of `Workload` → `PodGroup` → `Pod` creation.

### Non-Goals

- Multi-level / nested composite (`CompositePodGroup`) structures, since this KEP covers 
  single-level, flat `Job` workloads only.
- Implementing the integration in composite controllers (`JobSet`, `LWS`, `TrainJob`). Those 
  are pursued independently in their own repositories.
- Defining the `scheduling.k8s.io` building-block API or the `workloadbuilder` library 
  itself (owned by [KEP-6089] and consumed here).

## Proposal

This proposal builds on the recently introduced Workload-aware Scheduling enhancements. We assume
the reader is acquainted with the following KEPs:

- [KEP-4671]: Gang Scheduling.
- [KEP-5710]: Workload-aware preemption.
- [KEP-5732]: Topology-aware workload scheduling.
- [KEP-6089]: WAS Controller APIs.

The Job controller is extended to compile the user's scheduling intent into `Workload` and
`PodGroup` objects as part of its pod-management lifecycle, so that pods belonging to a Job are
scheduled according to the requested policy before they are created. The intent is expressed
through a new `spec.scheduling` field.

The key design principles for this alpha are:

> - One `Job` maps to one `Workload` with one `PodGroup` representing a single homogeneous group of

Reviewer 3: It's true for standalone Jobs.

If it comes with:
`scheduling.k8s.io/podgroup-template` and/or `scheduling.k8s.io/parent-composite-podgroup`
annotations - it may actually be part of a bigger structure.

Author: Updated.

  pods (a flat, single-level structure).
- The scheduling policy comes from the user's `spec.scheduling`, not from the Job's type. When

>   `spec.scheduling` is omitted, the controller defaults to the `Basic` policy.

Reviewer 3: Basic policy is just one thing.
It also defaults to Single disruption policy, no topology constraints and no pod-group RCs.

- Following the [KEP-6089], the controller always materializes a `Workload`/`PodGroup` for an 
  eligible Job, even for `Basic` scheduling policy.
- For `Gang`, an omitted `minCount` defaults to the Job's `parallelism` (or `completions` when
  parallelism is unset). `minCount` is mutable to support elastic scaling; all other
  `spec.scheduling` fields are immutable after creation.
- The Job controller does not create `Workload`/`PodGroup` when the Job carries an `OwnerReference`

>   to a parent controller that manages the `Workload` (e.g., `JobSet`). Such controllers set this

Reviewer 3: See my comment above - this is true for Workload.
But we may want job-controller still to manage (create) the PodGroup itself if appropriate annotation was set.
Or propagate the PodGroupName if the other annotation was set...

Author: Updated. PTAL.

scheduling constraints. This is a challenge for parallel applications (i.e., AI/ML training
workloads, MPI jobs) that require all pods to be scheduled and run together or none (gang
scheduling[^1]). The v1.36 alpha brought gang scheduling to the Job controller, but it did so by
inferring a hardcoded `Gang` policy from the Job's type rather than from explicit user intent.

Users have diverse use cases and require the ability to express explicit intent, such as opting
in or out of gang scheduling, requesting specific topologies, or configuring disruption policies
for their workloads. [KEP-6089] standardizes the reusable scheduling building blocks
(`scheduling.k8s.io/v1alpha3`) and a shared `workloadbuilder` translation library so that workload
controllers can expose these features consistently. This KEP integrates those building blocks into
the core `Job` API, "blazing the path" for the rest of the ecosystem while resolving the usability
gaps of the initial alpha.

### Goals

- Add a user-facing `spec.scheduling` (`JobSchedulingConfiguration`) field to the `batch/v1` Job,
  embedding the `scheduling.k8s.io/v1alpha3` building blocks (`policy`, `constraints`,

>   `disruptionMode`, `resourceClaims`) so users can express explicit scheduling intent.

Reviewer 5: Do we need to add anything related to Priority as we talked here: https://github.com/kubernetes/enhancements/pull/6092#discussion_r3403537954
cc @wojtek-t

Author: As been discussed in [KEP-6089](https://github.com/kubernetes/enhancements/pull/6092#discussion_r3403537954) and synced with @wojtek-t, we don't want to force users to set priority in different places (podTemplate and there) but we'd rather want controller to reconcile it once we support priority divergence within workload, but we're not there yet anyway. So, let's postpone it for now.

- Default to `Basic` scheduling when `spec.scheduling` is omitted, so the observable scheduling 
  outcome of existing Jobs is preserved (no all-or-nothing gate, and any number of schedulable 
  pods proceed to binding). Following the [KEP-6089], the controller still materializes a `Basic`
  `Workload`/`PodGroup`, which routes these pods through the Workload Scheduling Cycle 
  (batched scheduling and workload-aware preemption) without enforcing minCount.
- Let users opt in to `Gang` scheduling, with `minCount` defaulting to `parallelism` (or

>   `completions` when parallelism is unset) when omitted.

Reviewer 5: If parallelism is unset why minCount = completions ? IIUC, the default value for parallelism is 1, so it might be confusing why do we set minCount = completions

Author: Updated. PTAL.

- Compile `spec.scheduling` into `Workload`/`PodGroup` objects via the shared `workloadbuilder`
  library instead of bespoke controller logic.
- Support mutable `spec.scheduling.policy.gang.minCount` for elastic scaling, while keeping all
  other `spec.scheduling` fields immutable after creation. This relies on [KEP-4671] that makes 

>   `minCount` in `PodGroup`/`PodGroupTemplate` mutable in v1.37 to support workload scaling.

Reviewer 5: Are we planning to have this as part of 1.37?

Author: Yes (hopefully).

  the parent's compiled `Workload`/`CompositePodGroup`.
- The scheduling policy comes from the user's `spec.scheduling`, not from the Job's type. When
  `spec.scheduling` is omitted, the controller defaults to the `Basic` policy.
- Following the [KEP-6089], the controller always materializes a `Workload`/`PodGroup` for an 
  eligible Job, even for `Basic` scheduling policy.
- For `Gang`, an omitted `minCount` defaults to the Job's `parallelism` (or `completions` when
  parallelism is unset). `minCount` is mutable to support elastic scaling; all other
  `spec.scheduling` fields are immutable after creation.
- The Job controller does not create a `Workload` when the Job carries an `OwnerReference` to a
  parent controller that compiles and owns the `Workload` (e.g., `JobSet`). Such controllers set
  this `OwnerReference` when they create the Job. Whether the Job controller also skips `PodGroup`
  creation depends on what the parent delegates: if the parent injects the annotation 
  ([KEP-6089]), the Job controller still creates and manages the runtime `PodGroup` for its own 
  pods, mapping them to the parent's named `PodGroupTemplate` and attaching to the parent instance. 

>   If no such annotation is present, the parent owns both objects and the Job controller skips both.

Reviewer 5: Adding some thoughts on annotation here: https://github.com/kubernetes/enhancements/pull/6092#discussion_r3414343149

- Implementing the integration in composite controllers (`JobSet`, `LWS`, `TrainJob`). Those 
  are pursued independently in their own repositories.
- Defining the `scheduling.k8s.io` building-block API or the `workloadbuilder` library 
  itself (owned by [KEP-6089] and consumed here).

## Proposal

This proposal builds on the recently introduced Workload-aware Scheduling enhancements. We assume
the reader is acquainted with the following KEPs:

- [KEP-4671]: Gang Scheduling.
- [KEP-5710]: Workload-aware preemption.
- [KEP-5732]: Topology-aware workload scheduling.
- [KEP-6089]: WAS Controller APIs.

The Job controller is extended to compile the user's scheduling intent into `Workload` and
`PodGroup` objects as part of its pod-management lifecycle, so that pods belonging to a Job are
scheduled according to the requested policy before they are created. The intent is expressed
through a new `spec.scheduling` field.

The key design principles for this alpha are:

- One `Job` maps to one `PodGroup` representing a single group of pods. The `PodGroup` 
always links to a `Workload` via a `PodGroupTemplate`:
  * For a root Job the controller compiles the `Workload` itself

>   * For a non-root Job the `PodGroup` links to the parent-owned `Workload`

Reviewer 3: PodGroup should always link to Workload - for both non-root and root jobs.

Author: Updated. PTAL.

  name: <job-name>
  namespace: training
spec:
  parallelism: 4
  completions: 4
  completionMode: Indexed
  scheduling:               # New API field - scheduling intent
    policy:
      gang: {}              # minCount omitted -> defaults to parallelism (4)
    constraints:
      topology:
        - level: "topology.kubernetes.io/zone"
    disruptionMode:

>       all: {}               # entire group must be disrupted together

Reviewer 5: Do we have any user story when disruptionMode != all ?
If not, do we need to introduce this to the Job API in the first release?

Author: @wojtek-t @mm4tt , wdyt?

Reviewer 3: we definitely have - we need it for backward compatibility, that's the primary thing
but there are many examples of workloads, where I just want to paralellize stuff, but individual instances don't talk to each other and I rather prefer to run 3 instances if I can, than wait forever until I have a spot for all 100

      podGroupTemplateName: <podGroup-template-name>
  schedulingPolicy:
    gang:
      minCount: 4
```

> #### Example 2: Backward Compatibility and Sane Defaulting (Implicit Opt-Out)

Reviewer: Could we add another example for CronJob?

It seems we are covering this in the KEP but maybe pushing it up to example is good.

Author: Added. 
On this note, @wojtek-t , @mm4tt I would like to revisit this again. Does creating one`Workload`/`PodGroup` for each `Job` is the right call? Or should we tream the `CronJob` as a parent controller and create one `Workload` for the `CronJob` and multiple `PodGroup` for each `Job`?

Reviewer 2: I believe the answer to your question is below, each Job instance in a CronJob creates its own set of workload+podgroup. That seams reasonable, since each should be treated separately. This will especially important when you use the `Allow` as the concurrency policy for a cronjob, which means there can be MORE than a single job instance running at any point in time.

Reviewer 3: +1 to @soltysh 

It could be different if we would like to treat different incarnations of the Job as the same workload, but technically those are separate independent things.

        image: processor-image:v1
```

This compiles into a `Basic` scheduling policy:

```yaml
apiVersion: scheduling.k8s.io/v1alpha3
kind: Workload
metadata:
  name: <job-name>-<hash>
  namespace: batch
  ownerReferences:
  - apiVersion: batch/v1
    kind: Job
    name: <job-name>
    uid: <job-uid>
    controller: true
spec:
  controllerRef:
    apiVersion: batch/v1
    kind: Job
    name: <job-name>
  podGroupTemplates:
  - name: <podGroup-template-name>
    schedulingPolicy:
      basic: {}
---
apiVersion: scheduling.k8s.io/v1alpha3
kind: PodGroup
metadata:
  name: <workload-name>-<podGroup-template-name>-<hash>
  namespace: batch
  ownerReferences:
  - apiVersion: batch/v1
    kind: Job
    name: <job-name>
    uid: <job-uid>
    controller: true
  - apiVersion: scheduling.k8s.io/v1alpha3
    kind: Workload
    name: <workload-name>
    uid: <workload-uid>
spec:
  podGroupTemplateRef:
    workload:
      workloadName: <workload-name>
      podGroupTemplateName: <podGroup-template-name>
  schedulingPolicy:
    basic: {}
```

#### Example 3: CronJob with Gang Scheduling

A `CronJob` that periodically runs a gang-scheduled training Job. Each `Job` created by the

> `CronJob` is treated as standalone. `CronJob` is *not* a registered parent workload, so the Job

Reviewer 2: Can you define what a "registered parent workload" is? When I first read it, I thought it's any resource registered in kube-apiserver which sets `OwnerReference` field of a job. But, although CronJob matches those criteria you're calling it "not a registered". Why? That is not clear to me.

Author: Agree. Updated it across the KEP.

  or reject requests that populate both. The Job controller cooperates by skipping 
  `Workload`/`PodGroup` creation whenever the Job carries an `OwnerReference` to a registered 
  parent workload (replacing the v1.36 `spec.template.spec.schedulingGroup`-based opt-out).

- **Increased object count.** Because the controller now materializes a `Workload`/`PodGroup` for
  every eligible Job (Universal Representation), the number of scheduling objects grows relative to
  the v1.36 alpha, which only created objects for inferred gang Jobs. 
  * *Mitigation:* objects are small and garbage-collected with the Job; the Scalability section 
  quantifies the impact, and the feature stays behind the `WorkloadWithJob` feature gate for alpha.

- **Behavior change between alphas.** Jobs that were automatically gang-scheduled in v1.36 default
  to `Basic` in v1.37 unless the user sets `spec.scheduling.policy.gang`. 
  * *Mitigation:* gang is now an explicit opt-in; this is documented in the 
  [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy) and release notes.

- **Suspended Jobs and resource release** rely only on GC, which does not address releasing

>   resources (DRA) while a Job is suspended. This is acceptable for alpha. Future work may require

Reviewer: Can we make this a beta requirement?

I would like suspending jobs to cleanup workloads/PodGroup. One use case we have in Kueue is to relocate workloads to different racks upon suspend and in JobSet/TrainJob there is a interest in patching resources of a Suspended Job. This one shouldn't impact gangs but generally we have relaxed a lot of restrictions while a Job is suspended.

Author: It's already in the beta requirements. Reworded it, though.

        )
    }

    // 3. Create the flat, single-node logical tree for the Job (one PodGroup).
    rootNode := &workloadbuilder.WorkloadNode{
        Name:                "job-root",
        DefaultConfig:       defaultConfig,
        DefaultGangMinCount: ptr.To(job.Spec.Parallelism), // defaults to parallelism if MinCount is nil
        UserConfig:          userConfig,
    }

    // 4. Let the workloadbuilder compile and generate the Workload object.
    builder := workloadbuilder.NewBuilder(rootNode)
    workloadObj, err := builder.Build(
        context.Background(),
        job.Name,
        job.Namespace,
        metav1.NewControllerRef(job, jobKind),
    )
    if err != nil {
        return nil, err
    }

    return workloadObj, nil
}
```

#### API Validation via the `workloadbuilder` Library

`spec.scheduling` is validated in two layers with distinct responsibilities:

1. **api-server validation** owns the structural and *mutability* rules. It runs on every 
  create/update and must be self-contained (no dependency on cluster state or other 
  live objects). It checks:
   * exactly one scheduling policy is set (`basic` xor `gang`);

>    * `gang.minCount`, when set, is `>= 1`;

Reviewer: Should we validate that the `gang.minCount` can't exceed parallelism?

Author: I think we should not add a hard validation for `minCount > parallelism` for a few reasons:

- Both `minCount` and parallelism are independently mutable. If we reject `minCount > parallelism`, we force users to update parallelism before `minCount` when scaling up, which is fragile. For example, a controller doing elastic scaling would need to carefully order its two API calls.

- When `minCount` is nil it defaults to parallelism, and parallelism itself has its own defaulting. Crossfield validation on top of that gets messy and easy to get wrong.

- If `minCount > parallelism`, the gang can simply never be satisfied and the Job makes no progress, similar to requesting more resources than the cluster has. This is observable (pending pods, events) and self-correcting once the user fixes either value.

@wojtek-t wdyt?

Reviewer 3: I'm not sure I actually agree:

3) As you mentioned it can't really make progress - so by definition is a "faulty case".
 So if we can prevent that, we should actually do, as it's definitely not what users intended.

2) Defaulting always happens before validation. So validation happens on well defined state already.
I don't buy this argument at all.

1) This one I also don't buy. If I'm modifying `parallelism` as a Job owner, realistically I should change minCount in exactly the same APi call. I don't see any usecase where I would like to split those two into separate calls.
[Note that we're not talking about updates to Workload/PodGroup objects - it's purely about Job API.]

So I don't agree - I actually agree with Kevin that we should add such validation.

Reviewer: I guess the benefit of having an API to disable this is that in these more exotic cases someone could disable gang.

Like where completions > parallelism and I guess job users expect pods to finish at different times. In this case I think gang should just be disabled.

But I still think validation makes sense here.

Author: Updated. PTAL.

#### ML Training Job with Gang Scheduling

As a machine learning engineer, I want to run a distributed training job with 8 workers that must
all be scheduled together. I set `spec.scheduling.policy.gang` on the Job (optionally with a
topology constraint to co-locate the workers), so that if only 7 workers can be scheduled, no pods
start and no resources are wasted. I do not have to set `parallelism` and `completions`
in a specific way to "qualify" for gang scheduling; I declare my intent explicitly.

#### Standard Batch Job with Workload Tracking

As a data engineer, I want to run a batch processing job that processes files independently without

> gang scheduling requirements. I omit `spec.scheduling` entirely, so the Job defaults to `Basic`

Reviewer 3: I can also explicitly specify Basic policy to achieve a similar effect.

The core point in this user-story is backward compatibility - we should reflect that in its name.

Author: Updated. PTAL

(pod-by-pod) scheduling and behaves exactly as it does today. A `Basic` scheduling policy is
set for the Job, giving me a consistent objects to observe its scheduling state without
changing how the pods are scheduled.

### Notes/Constraints/Caveats

#### Alpha Constraints

- The alpha targets single-level `Job` workloads: one `Job` maps to one `Workload` with one `PodGroup`, 
  and all pods in the `Job` share a single scheduling policy. Elastic scaling is supported through the 
  mutable `gang.minCount`.
- `spec.scheduling.policy.gang.minCount` is mutable to support elastic scaling ([KEP-4671]); all 
  other `spec.scheduling` fields are immutable after creation.
- Opting out of WAS object creation happens implicitly in two ways: omitting `spec.scheduling`
  yields `Basic` (the controller still owns the objects), and a Job whose parent controller manages

>   the `Workload` is skipped entirely.

Reviewer 3: Same comment about jobs having ownerRefs I had above.

Reviewer 3: And there are more places below...

Author: Addressed all. PTAL.

  or reject requests that populate both. The Job controller cooperates by skipping 
  `Workload`/`PodGroup` creation whenever the Job carries an `OwnerReference` to a registered 
  parent workload (replacing the v1.36 `spec.template.spec.schedulingGroup`-based opt-out).

- **Increased object count.** Because the controller now materializes a `Workload`/`PodGroup` for
  every eligible Job (Universal Representation), the number of scheduling objects grows relative to
  the v1.36 alpha, which only created objects for inferred gang Jobs. 
  * *Mitigation:* objects are small and garbage-collected with the Job; the Scalability section 
  quantifies the impact, and the feature stays behind the `WorkloadWithJob` feature gate for alpha.

- **Behavior change between alphas.** Jobs that were automatically gang-scheduled in v1.36 default
  to `Basic` in v1.37 unless the user sets `spec.scheduling.policy.gang`. 
  * *Mitigation:* gang is now an explicit opt-in; this is documented in the 
  [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy) and release notes.

- **Suspended Jobs and resource release** rely only on GC, which does not address releasing

>   resources (DRA) while a Job is suspended. This is acceptable for alpha. Future work may require

Reviewer 3: nit: please reword, as we already have beta requirement that we will delete PodGroup on suspend (i.e. it's not "may").

  the controller to delete `PodGroup`/`Workload` on suspend and recreate on resume.

## Design Details

### Core Principles & Assumptions

The integration follows the Workload-aware Scheduling design principles from [KEP-6089] for the 
single-level `Job`:

- **The Root Controller is the Compiler.** For a standalone `Job`, the Job controller is the
  root-most controller and is responsible for compiling, creating, and managing the

>   scheduler-facing `Workload` and its runtime `PodGroup`. When a `Job` instead carries an

Reviewer 3: Yes for Workload. Not necessarily for PodGroup.

Author: Reworded.

  `OwnerReference` to a parent controller that manages the `Workload` (e.g., a `JobSet` owns the Job),
  the Job controller observes that linkage and *bypasses* creating any `Workload`, so the parent
  remains the single source of truth.
- **Separation of Structure and Policy.** The `Job` spec owns the workload *structure*

>   (`parallelism`, `completions`, the pod template); the user owns the scheduling *policy* via

Reviewer 3: I don't understand it - scheduling policy is also part of JobSpec.
They are both owned by the user.

Author: I think it was bad wording. I removed it.

  `spec.scheduling`. The Job controller acts as the translator that combines the two and compiles
  the low-level `Workload`/`PodGroup` for the scheduler.
- **Universal Representation.** Standard pod-by-pod scheduling is a first-class policy (`Basic`).
  The controller always emits a `Workload`/`PodGroup` for an eligible Job, using `Basic` as the
  backward-compatible default. `Basic` keeps the standard scheduling outcome and does not impose
  all-or-nothing semantics.
- **Sane Defaults and Escape Hatches.** A `Job` defaults to `Basic`. Users can opt in to `Gang`,
  request topology constraints, and configure disruption modes; an escape hatch lets a user request

>   topology-only co-location while keeping the `Basic` policy.

Reviewer 3: topology contraints are not well defined behavior with Basic policy.
[In theory we can schedule first pod and may not be able to schedule anything else later.]

We should not allow for it, at least as a starting point (we can always relax it later if needed).

Author: Removed the topology constraints from here.

field's defaulting, validation, and controller compilation only run when the gate is enabled.

This typed, user-facing field replaces the v1.36 alpha's implicit mechanisms: the type-based
automatic policy inference and the `spec.template.spec.schedulingGroup`-based opt-out are no longer
how users express or suppress scheduling intent.

#### Go Package Placement & Graduation

Embedding a pre-stable `scheduling.k8s.io/v1alpha3` type inside the GA `batch/v1.JobSpec` is
permitted while the `Scheduling` field itself remains alpha-gated. Following the transition pattern
described in [KEP-6089], when the field graduates to default-enabled the building blocks graduate
straight into the stable `scheduling.k8s.io/v1` package, the `batch/v1.JobSpec` field is updated to
reference the `v1` type, and Go type aliases are left in the `v1alpha3` package so third-party
controllers that still import the alpha package continue to compile.

### Integration with the workloadbuilder Library

The Job controller compiles `spec.scheduling` into a `Workload` using the `workloadbuilder` library.

#### Library Dependency and Packaging

> The `workloadbuilder` library lives alongside the `scheduling.k8s.io/v1alpha3` types so it 

Reviewer 3: It doesn't leave alongside the types. Type live in the API staging repo.

The workload builder will leave in components helper...

Author: Updated.

can be vendored by both in-tree controllers (the Job controller, here) and out-of-tree 
controllers. The Job controller depends on it as a normal package import; this KEP does 
not define or fork the library, it only consumes its `NewBuilder`/`Build`, `WorkloadNode`, 
and `MapPodGroupConfig` surface. If the library API shifts before it stabilizes, the Job 
integration tracks those changes through the shared dependency rather than maintaining 
its own copy.

#### Division of Responsibilities

> The boundary between the library and the Job controller is explicit.

Reviewer 3: I personally don't see any value in this subsection - this is exactly how the library was defined in the other KEP and it kind of repeats it.

For the sake of shortening the (already pretty long) KEP, I would just remove this part.

Reviewer 2: Agree with Wojtek here, especially that the next section covers the implementation bit, thus covering the integration details.

Author: Removed.

   guarantees that any configuration the API server accepts is one the controller can compile into a
   valid `Workload`, rejecting combinations that are structurally legal but semantically invalid
   (e.g. an unsupported disruption mode, or a policy/constraint pairing the builder cannot translate)
   uniformly across all integrating controllers. Because resolution reads only the incoming object
   and never cluster state, it is safe to call from the registry layer; it complements, and does not
   replace, the structural and immutability checks in the api-server validation.

#### Instantiating the runtime `PodGroup`

`Build` returns only the `Workload` (the scheduling template); it does not create the runtime
`PodGroup`. After the `Workload` exists on the API server, the Job controller instantiates the
`PodGroup` from the `Workload`'s single `PodGroupTemplate`. For a Job there is exactly one template,
so the controller creates one `PodGroup` that references the template and carries two
ownerReferences — a controller ref to the `Job` (so it is GC'd with the Job) and a non-controller
ref to the `Workload`:

> ```go

Reviewer 3: Do we really need this code?

Author: Removed.

            len(workload.Spec.PodGroups))
    }
    tmpl := workload.Spec.PodGroups[0]

    pg := &schedulingv1alpha3.PodGroup{
        ObjectMeta: metav1.ObjectMeta{
            Name:      podGroupName(workload.Name, tmpl.Name),
            Namespace: job.Namespace,
            OwnerReferences: []metav1.OwnerReference{
                *metav1.NewControllerRef(job, jobKind),         // controller: true
                workloadOwnerRef(workload),                     // controller: false
            },
        },
        Spec: schedulingv1alpha3.PodGroupSpec{
            // Link the runtime instance back to its template in the Workload.
            PodGroupTemplateReference: schedulingv1alpha3.PodGroupTemplateReference{
                Workload: schedulingv1alpha3.WorkloadReference{WorkloadName: workload.Name},
                Name:     tmpl.Name,
            },
        },
    }
    return pg, nil
}
```

Pods are then created by the existing Job pod-management logic with
`pod.Spec.SchedulingGroup.PodGroupName` set to the `PodGroup`'s name, which is what the scheduler
keys on for gang/topology behavior.

> 

Reviewer 3: @wojtek-t : Note to myself - continue from here :)

topology constraint to co-locate the workers), so that if only 7 workers can be scheduled, no pods
start and no resources are wasted. I do not have to set `parallelism` and `completions`
in a specific way to "qualify" for gang scheduling; I declare my intent explicitly.

#### Standard Batch Job with Workload Tracking

As a data engineer, I want to run a batch processing job that processes files independently without
gang scheduling requirements. I omit `spec.scheduling` entirely, so the Job defaults to `Basic`
(pod-by-pod) scheduling and behaves exactly as it does today. A `Basic` scheduling policy is
set for the Job, giving me a consistent objects to observe its scheduling state without
changing how the pods are scheduled.

### Notes/Constraints/Caveats

#### Alpha Constraints

- The alpha targets single-level `Job` workloads: one `Job` maps to one `Workload` with one `PodGroup`, 
  and all pods in the `Job` share a single scheduling policy. Elastic scaling is supported through the 
  mutable `gang.minCount`.
- `spec.scheduling.policy.gang.minCount` is mutable to support elastic scaling ([KEP-4671]); all 
  other `spec.scheduling` fields are immutable after creation.

> - Opting out of WAS object creation happens implicitly in two ways: omitting `spec.scheduling`

Reviewer 2: That's not quite true, is it? Leaving empty `.spec.scheduling` still creates workload+podgroup pair, but those will only produce the basic, ie. pod-by-pod scheduling. So, there's NO way to opt out of WAS object creations, right?

Author: This is tricky, since we consider not choosing `Gang` policy as opting out. But is that in fact an opt-out? Should we call it opt-out while we are creating a `Workload`/`PodGroup` with a `Basic` policy? 

@wojtek-t , wdyt?

Reviewer 3: The primary question is: what do you want to achieve by opt-out.

I think there are two reasons:
1) There is a bug in this integration - for that we have FG and we should simply disable it in Alpha/Beta release.
2) You don't want to have the integration on - I see two subcases:
a) you want to preserve the original behavior - for that the default behavior should achieve it (so not setting .spec.scheduling), because Basic policy will give you exactly that
b) you don't want the Workload/PodGroup object - that I don't think it's a reasonable request

So I personally think we don't need anything else because of ^^.

@soltysh - do you see a problem with that or your comment was primarily a wording comment?
[Because I technically agree that there is not way of opt-out if we read that literally.]

Reviewer 2: Mostly wording, I don't see a problem with empty `.spec.scheduling` creating the workload+podgroup pair for pod-by-pod scheduling. I just want to make sure that other readers of this document have a clear understanding.

Author: Updated it across the KEP. PTAL.

straight into the stable `scheduling.k8s.io/v1` package, the `batch/v1.JobSpec` field is updated to
reference the `v1` type, and Go type aliases are left in the `v1alpha3` package so third-party
controllers that still import the alpha package continue to compile.

### Integration with the workloadbuilder Library

The Job controller compiles `spec.scheduling` into a `Workload` using the `workloadbuilder` library.

#### Library Dependency and Packaging

The `workloadbuilder` library lives alongside the `scheduling.k8s.io/v1alpha3` types so it 
can be vendored by both in-tree controllers (the Job controller, here) and out-of-tree 
controllers. The Job controller depends on it as a normal package import; this KEP does 
not define or fork the library, it only consumes its `NewBuilder`/`Build`, `WorkloadNode`, 
and `MapPodGroupConfig` surface. If the library API shifts before it stabilizes, the Job 
integration tracks those changes through the shared dependency rather than maintaining 
its own copy.

#### Division of Responsibilities

The boundary between the library and the Job controller is explicit.

The `workloadbuilder` library is responsible for:

>   * Defining the polymorphic, hierarchy-agnostic IR (`SchedulingConfig`, `WorkloadNode`) that 

Reviewer 2: What the IR stands for here? Is it intermediate representation or something else? Let's use a non-abbreviated version and introduce the acronym next to it, so that every reader of this document can fully understand what is being discussed.

Author: Updated.

  `spec.scheduling`.
  * Emitting the `scheduling.k8s.io/v1alpha3` `Workload` object (the scheduling template, including its
  `PodGroupTemplate`) via `Build`.

The Job controller is responsible for:

  * Choosing the batch-specific defaults — `Basic` scheduling and a fallback gang size of
  `spec.parallelism` and passing them to the library.
  * Mapping `spec.scheduling` into the library IR with the leaf helper `MapPodGroupConfig`.
  * Instantiating the runtime `PodGroup` from the compiled `Workload`, since `Build` only produces the
    `Workload` template.
  * Discovery, ownerReferences, garbage collection, and object creation ordering, as described in
  [Job Controller Changes](#job-controller-changes).

Because a standalone `Job` is a single-level workload, the controller builds a **flat, single-node**
`WorkloadNode` tree (`len(Children) == 0`) and uses the leaf mapping helper (`MapPodGroupConfig`); it
never uses `MapCompositeGroupConfig`, which is reserved for multi-level controllers.

#### Building the Logical Tree and Compiling the `Workload`

The controller's `generateWorkload` helper performs four steps: 

>   1. Set the Job's domain default to `Basic`.

Reviewer 2: ```suggestion
  1. Set the Job's default configuration to `Basic`.
```
?

        },
    }

    // 2. Map the public Job.Spec.Scheduling wrapper into the library's IR via the leaf helper.
    var userConfig *workloadbuilder.SchedulingConfig
    if job.Spec.Scheduling != nil {
        userConfig = workloadbuilder.MapPodGroupConfig(
            job.Spec.Scheduling.Policy,
            job.Spec.Scheduling.Constraints,
            job.Spec.Scheduling.DisruptionMode,
            job.Spec.Scheduling.ResourceClaims,
        )
    }

    // 3. Create the flat, single-node logical tree for the Job (one PodGroup).
    rootNode := &workloadbuilder.WorkloadNode{
        Name:                "job-root",
        DefaultConfig:       defaultConfig,
        DefaultGangMinCount: ptr.To(job.Spec.Parallelism), // defaults to parallelism if MinCount is nil
        UserConfig:          userConfig,
    }

    // 4. Let the workloadbuilder compile and generate the Workload object.
    builder := workloadbuilder.NewBuilder(rootNode)
    workloadObj, err := builder.Build(
        context.Background(),
        job.Name,
        job.Namespace,
        metav1.NewControllerRef(job, jobKind),
    )
    if err != nil {
        return nil, err
    }

    return workloadObj, nil
}

> ```

Reviewer 2: I'm not a fan of exact code bits in the documents, they tend to age poorly. Rather it's better to have a high-level pseudo code, which outlines just the logical flow. So here, I'd drop imports - not relevant, and error checks. Just focus on the main bits of generating the workload.

Author: Updated. PTAL
We have this in [KEP-6089](#6092), do we need to update the code there as well? Maybe it's ok there as an example.

Reviewer 3: Given this code is obvious enough (based on the text above it), I would actually remove this code completely. It doesn't introduce additional value on top of what we already have described in the parent KEP...

Author: Removed.

start and no resources are wasted. I do not have to set `parallelism` and `completions`
in a specific way to "qualify" for gang scheduling; I declare my intent explicitly.

#### Backward-Compatible Standard Batch Job

As a data engineer, I want to run a batch processing job that processes files independently without
gang scheduling requirements. I omit `spec.scheduling` entirely (or set `spec.scheduling.policy.basic`
explicitly for the same effect), so the Job defaults to `Basic` scheduling. The observable scheduling
outcome matches a standard Job, while a `Basic` `Workload`/`PodGroup` is still materialized, giving me consistent objects to observe its scheduling state.

### Notes/Constraints/Caveats

#### Alpha Constraints

- The alpha targets single-level `Job` workloads: one `Job` maps to one `PodGroup`, and all pods in
  the `Job` share a single scheduling policy. Elastic scaling is supported through the mutable `gang.minCount`.
- `spec.scheduling.policy.gang.minCount` is mutable to support elastic scaling ([KEP-4671]); all 
  other `spec.scheduling` fields are immutable after creation.

> - Opting out of `Workload` creation happens implicitly in two ways: omitting `spec.scheduling`

Reviewer 3: This is misleading - it's not really opting out of "Workload" creation. 
Technically you don't have a way to opt-out from it.

We should differentiate two things:
1) the Workload/PodGroup creation - there is technically no way to opt-out from this feature (other than initially disabling FG)
2) different scheduling behavior (e.g. gang scheduling) - this is an explicit opt-in via `spec.scheduling` - so by default as end-user you get the original behavior (even though we still create Workload/PodGroup in such case)

Author: Updated. PTAL

reference the `v1` type, and Go type aliases are left in the `v1alpha3` package so third-party
controllers that still import the alpha package continue to compile.

### Integration with the workloadbuilder Library

The Job controller compiles `spec.scheduling` into a `Workload` using the `workloadbuilder` library.

#### Library Dependency and Packaging

The `scheduling.k8s.io/v1alpha3` building-block types live in the API staging repo
(`k8s.io/api/scheduling/v1alpha3`). The `workloadbuilder` library lives separately in
`k8s.io/component-helpers`, so it can be vendored by both in-tree controllers and out-of-tree 
controllers. The Job controller consumes its `NewBuilder`/`Build`, `WorkloadNode`, and 
`MapPodGroupConfig`. If the library API shifts before it stabilizes, the Job integration 
tracks those changes through the shared dependency rather than maintaining its own copy.

#### Building the Logical Tree and Compiling the `Workload`

The controller's `generateWorkload` helper performs four steps: 
  1. Set the Job's default configuration to `Basic`.
  2. Map the user-facing `spec.scheduling` block into the library IR via `MapPodGroupConfig`.

>   3. Assemble a single-node `WorkloadNode`, supplying `DefaultGangMinCount` from `spec.parallelism` 

Reviewer 3: nit: depends on the discussion how we will pass such parameters
But this comment doesn't really block this KEP, we can update it later.

Reviewer 3: actually let's rephrase it to:

Assemble a single-node `WorkloadeNode`, if needed supplying minCount for gang from `spec.parallelism`.

Author: updated.

`spec.scheduling` is validated in two layers with distinct responsibilities:

1. **api-server validation** owns the structural and *mutability* rules. It runs on every 
  create/update and must be self-contained (no dependency on cluster state or other 
  live objects). It checks:
   * exactly one scheduling policy is set (`basic` xor `gang`).
   * `gang.minCount`, when set, is `>= 1`and does not exceed `spec.parallelism`. A gang 
   larger than the pod count can never be satisfied and the Job would stall indefinitely 
   with pending pods, so this faulty state is rejected at admission rather than left to 
   surface only at runtime. An elastic scale-up that sets `spec.parallelism` and 
   `gang.minCount` in the same request is validated against the final state.
   * topology constraints, disruption mode, and resourceClaims are individually well-formed;
   * on update, every `spec.scheduling` field is immutable **except** `gang.minCount`.
2. **`workloadbuilder` semantic validation** owns the *consistency* rules. The API server calls the
   library's `Validate` entrypoint, which performs the same configuration resolution and policy

>    validation that `Build` runs — resolving `DefaultConfig` + `DefaultGangMinCount` + `UserConfig`

Reviewer 3: I suggest removing `DefaultGangMinCount` from here - this is a detail that may change.

   - **Annotation present (PodGroup delegated):** the parent owns the `Workload` but expects the Job
     to manage its own runtime `PodGroup`. Proceed to step 5, creating the `PodGroup` mapped to the
     parent's named `PodGroupTemplate` (annotation value) and attaching it to the parent instance
     named by `scheduling.k8s.io/parent-composite-podgroup`. The `PodGroup` gets a controller
     `ownerReference` to the Job.
   - **Annotation absent (both delegated):** the parent owns both the `Workload` and the `PodGroup`.
     The Job controller skips creation entirely, discovers any existing objects, and uses them when
     creating pods.
2. If the Job already has pods (active or terminal pods owned by this Job), skip creation and only
   discover existing objects.
3. Look up existing `Workload`(s) in the Job's namespace whose `spec.controllerRef` points to this
   Job. If the `Workload` was created by the Job controller, it also has a controller
   `ownerReference` pointing to this Job (`controller: true`).
  - If none found, compile a `Workload` from the Job's `spec.scheduling` and create it with an 
  `ownerReference` and `spec.controllerRef` pointing to this Job.

>   - If more than one, treat as ambiguous and fall back (update a condition or trigger an event).

Reviewer 3: Fallback sounds fine for Alpha.

But I suggest adding explicit Beta graduation criteria to revisit that - I think that for Beta we should consider halting processing of that Job and exposing the "mallformed state" via condition.

Reviewer 3: Same for PodGroups.

Reviewer 3: @soltysh - for your thoughts too

Reviewer 2: Agree with what Wojtek proposes here. This way we'll match what is written in `Reconcile Integration and Error Handling` section, regarding error handling from the workloadbuilder library. I'd treat this situation in the exact same way.

Author: Updated beta criteria. PTAL.

#### Controller Workflow

The Job controller attempts to create `Workload` and `PodGroup` only when the Job has no pods associated with it 
(no active or terminal pods owned by the `Job`). If the Job already has one or more pods, the controller only 
discovers and uses existing `Workload`/`PodGroup` if any and does not create new ones. This rule is important for 
correctness when the controller restarts or is upgraded in the middle of the workflow (i.e., after creating 
`Workload` but before creating `PodGroup` or pods). On the next sync, the controller will find the existing objects 
via informers/listers and continue.

The controller discovers or creates `Workload` and `PodGroup` as follows:

1. If the Job carries an `OwnerReference` to a parent controller that manages the `Workload`
   (i.e., a higher-level controller such as `JobSet` that compiles and owns the workload tree),

>    skip `Workload`/`PodGroup` creation entirely since the parent owns them. The controller still

Reviewer 3: Depending on the discussion in: https://github.com/kubernetes/enhancements/pull/6153/changes#r3388799503
we may need to adjust it.

    Pod -->|ownerRef| PodGroup
```

- The `Workload` object has an ownerReference to the `Job` object with `controller: true` in case 
  it was created by the Job controller.
- The `PodGroup` object has an ownerReference to the `Job` object with `controller: true` in case 
  it was created by the Job controller and another ownerReference to the `Workload` object.
- The `Pod` object has an ownerReference to the `Job` object with `controller: true` and another 
  ownerReference to the `PodGroup` object

By this ownerReferences relationship, garbage collection will remove objects accordingly that avoids orphaned Pods with a stale PodGroup reference.

#### Defaulting Rules

To ensure backward compatibility and avoid breaking existing pipelines:

- If `Scheduling` is unset or `Policy` is nil, the Job controller defaults to the `Basic` scheduling
  policy (standard pod-by-pod scheduling).
- If a user selects `Gang` but leaves `MinCount` unset, the controller injects a context-aware sane
  default of `MinCount = parallelism` (or `MinCount = completions` when `parallelism` is unset).

> - Users can configure an escape hatch — for example, requesting only topology constraints while

Reviewer 3: Topology without gang behavior in general has weird corner cases. So let's remove this point to not steer people in that direction...

Author: Removed.

    Pod -->|ownerRef| PodGroup
```

- The `Workload` object has an ownerReference to the `Job` object with `controller: true` in case 
  it was created by the Job controller.
- The `PodGroup` object has an ownerReference to the `Job` object with `controller: true` in case 
  it was created by the Job controller and another ownerReference to the `Workload` object.
- The `Pod` object has an ownerReference to the `Job` object with `controller: true` and another 
  ownerReference to the `PodGroup` object

By this ownerReferences relationship, garbage collection will remove objects accordingly that avoids orphaned Pods with a stale PodGroup reference.

#### Defaulting Rules

> - **`Scheduling` unset → `Basic`.** Existing Jobs and users that do not opt into WAS carry no 

Reviewer 3: I think this comment got lost:
https://github.com/kubernetes/enhancements/pull/6153#discussion_r3388761465

  it was created by the Job controller.
- The `PodGroup` object has an ownerReference to the `Job` object with `controller: true` in case 
  it was created by the Job controller and another ownerReference to the `Workload` object.
- The `Pod` object has an ownerReference to the `Job` object with `controller: true` and another 
  ownerReference to the `PodGroup` object

By this ownerReferences relationship, garbage collection will remove objects accordingly that avoids orphaned Pods with a stale PodGroup reference.

#### Defaulting Rules

These rules are applied by the controller when it compiles the
`Workload`/`PodGroup`; they are not api-server field defaulting. 
This is a controller-side resolution that is required to resolve the unset case anyway.

- **`Scheduling` unset → `Basic`.** Existing and non-WAS Jobs carry no `spec.scheduling`, the
  controller resolves the absent policy to `Basic`, preserving their behavior.
- **`Scheduling` set but `Policy` nil → `Basic`.** `WorkloadPodGroupSchedulingPolicy` is a 
  discriminated union for which the compiled `PodGroup` must carry exactly one concrete policy, so a
  nil `Policy` is resolved to `Basic`.

> - **`Gang` with `MinCount` unset → `MinCount = parallelism`.** a context-aware sane gang size supplied via `DefaultGangMinCount`.

Reviewer 3: Let's remove the "supplied via DefaultGangMinCount" - this is implementation detail and I still don't think it should be done this way :)

Also - let's clarify that this will defaulting will be done also controller-side (otherwise we wouldn't have a way to dstinguish later (on update) from user-provided one).

Author: Updated. PTAL

    Pod -->|ownerRef| PodGroup
```

- The `Workload` object has an ownerReference to the `Job` object with `controller: true` in case 
  it was created by the Job controller.
- The `PodGroup` object has an ownerReference to the `Job` object with `controller: true` in case 
  it was created by the Job controller and another ownerReference to the `Workload` object.
- The `Pod` object has an ownerReference to the `Job` object with `controller: true` and another 
  ownerReference to the `PodGroup` object

By this ownerReferences relationship, garbage collection will remove objects accordingly that avoids orphaned Pods with a stale PodGroup reference.

#### Defaulting Rules

To ensure backward compatibility and avoid breaking existing pipelines:

- If `Scheduling` is unset or `Policy` is nil, the Job controller defaults to the `Basic` scheduling

>   policy (standard pod-by-pod scheduling).

Reviewer 4: Wondering why do we need to default on a nil `Policy`? Is expecting `Policy` to be set a backward-incompatible rule or breaking existing pipelines (according to the sentence above)?

Reviewer 3: +1 to this question

On a related note - if we're defaulting the Policy, why we're not defaulting the DisruptionMode for example?

Author: > Wondering why do we need to default on a nil Policy? Is expecting Policy to be set a backward-incompatible rule or breaking existing pipelines (according to the sentence above)?

These were two separate things bundled together (I split them in the KEP):
- Scheduling unset → Basic: this is the actual backward-compatibility case (existing Jobs have no `spec.scheduling`).
- Scheduling set, Policy nil → Basic: (not a compatibility concern) You're right that requiring Policy here wouldn't break anything. We default it only because `WorkloadPodGroupSchedulingPolicy` is a required union, so nil must resolve to something. If we'd prefer explicit intent, we could instead require (validation) Policy when Scheduling is set, fine either way.

> On a related note - if we're defaulting the Policy, why we're not defaulting the DisruptionMode for example?

Policy is required, `DisruptionMode` is optional (same for Constraints = no topology, ResourceClaims = none).

Reviewer 3: > Scheduling unset → Basic: this is the actual backward-compatibility case (existing Jobs have no spec.scheduling).

We need such behavior.
But technically, we don't need to default it explicitly and can handle unset case in the controller.
So the question is why do you need to default it.

The potential answer could be that we want to avoid explicit defaultin on the controller side, but unfortunately we need it anyway.
If there is a Job that has no pods on upgrade, it will not have the fields set and the controller should still handle it.

So if we need to handle that anyway on controller side, I don't see good reason why we need to explicitly default it.

Reviewer 3: /cc @soltysh

Reviewer 2: Agree with Wojtek here. The only reason for defaulting I can think of is informational.  But that can be done through explicit description of the field, which is required.

Author: Updated. PTAL.

#### Handling Updates and Mutability

To support dynamic scaling of gang-scheduled workloads (Elastic Jobs), the Job API allows in-flight
updates to `spec.scheduling.policy.gang.minCount`; all other `spec.scheduling` fields are immutable
after Job creation, and updates that change them are rejected by API validation. This replaces the
v1.36 validation that rejected `spec.parallelism` updates for gang Jobs: because the gang size is
now driven by the mutable `minCount` ([KEP-4671]), `spec.parallelism` is no longer frozen for gang
Jobs, restoring support for [Elastic Indexed
Jobs](https://kubernetes.io/docs/concepts/workloads/controllers/job/#elastic-indexed-jobs). API
validation reuses the `workloadbuilder` library where possible so the accepted configurations stay
consistent with what the controller actually compiles. The update-validation rules change as follows:

  * `spec.parallelism` becomes *mutable* again: the v1.36 rule that rejected `spec.parallelism`
  updates for gang Jobs is removed, restoring Elastic Indexed Jobs.
  * `spec.scheduling.policy.gang.minCount` is *mutable* in-flight: on change the controller

>   recompiles the `Workload` and re-syncs the `PodGroup` size.

Reviewer 4: IIRC in the document proposing minCount mutability we decided **not to** update BYO PodGroup's `minCount` when the Job spec changes. Could you reflect that in the KEP?

Reviewer 4: Also, in general, how the new `spec.scheduling` fields will work with the BYO Workload/PodGroup?

Author: Added a dedication section for BYO. PTAL.

  - Truncation of workload name and podGroup name is applied when necessary to respect name length limits.
  - The hash allows multiple PodGroups within a `Workload` and `PodGroupTemplate` to have distinct names. For alpha, the controller creates a single `PodGroup` per Job, however, the pattern still supports future multi-PodGroup cases.

### Deletion and Garbage Collection

The Job controller does not explicitly delete `Workload` or `PodGroup`. However, in the case of the controller
creating them, it sets `ownerReferences` so that garbage collection removes them when the Job is deleted.
No additional controller logic is required for deletion in the current design.

The Job controller does not add or adopt ownerReferences on objects it did not create (user-created or higher-level controller-created objects). Users or other controllers may create Workloads/PodGroups with the same ownerReferences as the Job controller would use.

To distinguish controller-created objects from user-created ones that may have the same ownerReferences,

> the Job controller may set a `managed-by` annotation or equivalent metadata on `Workload` and `PodGroup` objects it creates.

Reviewer 3: IIRC we didn't implement that for 1.36. Do we plan to implement that for 1.37?

This is also relevant to the question here: https://github.com/kubernetes/enhancements/pull/6153/changes#r3388896561

Author: I'd include it for 1.37 to enable better integration with JobSet.

- Passthrough: topology constraints, disruption mode, and resourceClaims declared in
  `spec.scheduling` appear in the compiled `Workload`/`PodGroup`.
- Failure Recovery test (create a Job while the `Workload` API is unavailable, verify the controller
  retries and the `Workload` is eventually created).
- Feature gate disable/enable (Jobs work without `Workload`/`PodGroup` creation; Jobs with an
  `OwnerReference` to a parent workload do not create `Workload`/`PodGroup`).
- Jobs created by CronJob get one `Workload` and one `PodGroup` per Job, and these are GC'd when the
  Job completes or is deleted.
- When a Job is suspended, pods are deleted but `Workload`/`PodGroup` remain; on resume the same
  `Workload`/`PodGroup` are used and pods are recreated with the correct `schedulingGroup`.
- Verify controller-created `Workload`/`PodGroup` have the correct owner references.

##### e2e tests

- End-to-end gang scheduling: all pods are scheduled together or none.
- `Basic` Jobs schedule pod-by-pod while still producing a `Workload`/`PodGroup`.
- Elastic gang resize via `minCount` update.
- Mixed workloads: gang and basic Jobs coexist.

> - Failure scenarios, e.g., insufficient resources for a gang, partial failures.

Reviewer: Should we add a e2e test for CronJob?

Author: Updated the e2e.

- Non-gang scheduling Jobs do not have `Workload`/`PodGroup` objects created.
- Jobs managed by higher-level controllers skip `Workload`/`PodGroup` creation.
- API validation rejects updates that change `spec.parallelism` for gang scheduling Jobs.
- Unit and integration tests for the `Workload`/`PodGroup` creation flow.

#### Alpha (v1.37)

This second alpha replaces the automatic model with the user-facing API:
- New `spec.scheduling` (`JobSchedulingConfiguration`) field added to `batch/v1`, gated by the
  existing `WorkloadWithJob` feature gate (still default-disabled).
- The Job controller compiles `spec.scheduling` into `Workload`/`PodGroup` via the shared
  `workloadbuilder` library, defaulting to `Basic` and materializing a `Workload`/`PodGroup` for
  every eligible Job (Universal Representation).
- `Gang` opt-in with `minCount` defaulting to `parallelism`, plus support for mutable `minCount`
  (elastic scaling) and passthrough of topology constraints, disruption mode, and resourceClaims.
- API validation makes `spec.scheduling` fields immutable except `gang.minCount`; the v1.36
  `spec.parallelism`-rejection validation is removed.

> - Jobs owned by a higher-level controller (via `OwnerReference`) skip `Workload`/`PodGroup` creation.

Reviewer 3: Same here - pending discussion in https://github.com/kubernetes/enhancements/pull/6153/changes#r3388799503

- Unit and integration tests for the new API, defaulting, mutability, and `workloadbuilder`
  compilation; user-facing documentation for the new API.

#### Beta

- Promote the `WorkloadWithJob` feature gate to enabled by default.
- Evaluate whether the Job controller's [current batch-create](https://github.com/kubernetes/kubernetes/blob/2023f445eca52e6baa72139e56c6e4e01be0ee97/pkg/controller/job/job_controller.go#L1780-L1838) of pods should change when gang scheduling is active (it slows pod creation), and document the decision.

> - The controller must delete `PodGroup`/`Workload` when a Job is suspended and recreate them on

Reviewer 3: Should we delete both PodGroup and Workload, or just PodGroup?

What we really need is to free-up the resources, so the hard requirement is about PodGroup only imho.

Also - if the Workload is managed by higher-level controller, it will probably not delete it either on Job suspension. So keeping the workload on suspension makes it more similar to that usecase.

@kannon92 - for you thoughts too

Reviewer: I think the main reason to delete PodGroup and Workload is that the job api spec can change during suspend and resume.

So right now, we can get disconnects between Workload minCount and PodGroup. I think its best to delete both and recreate the objects on resume.

Reviewer 3: > I think the main reason to delete PodGroup and Workload is that the job api spec can change during suspend and resume.

Do we allow arbitrary changes? I might be missing something but I think this won't influence the shape of the workload itself, no?

/cc @soltysh

Reviewer 2: We only allow a limited set of changes to a job, such as [scheduling directives](https://kubernetes.io/docs/concepts/workloads/controllers/job/#mutable-scheduling-directives) and [pod resources](https://kubernetes.io/docs/concepts/workloads/controllers/job/#mutable-pod-resources-for-suspended-jobs) none of which will affect Workload. So it should be sufficient to only delete the PodGroup.

Reviewer 5: If users modify scheduling directives, they might also want to change the constraints, isn't?
I think we may want to either:
1. Delete the Workload and PodGroup objects when the Job is suspended, or
2. Make these fields mutable in the future, similar to what we did for `minCount`.

In the future, we might have more use-cases when Job spec changes are required during the suspend/resume: https://github.com/kubeflow/trainer/pull/3559

cc @tenzen-y @VassilisVassiliadis @astefanutti

Author: As per our offline sync, we will decide for beta which objects should be removed.PTAL

### Upgrade / Downgrade Strategy

- **Upgrade:**

>   1. Upgrade kube-apiserver first, so it can serve `scheduling.k8s.io/v1alpha3` and accept 

Reviewer 3: nit: we will hopefully migrate to v1beta1 in 1.37

Author: We can fix it later if this hopefully happens.

    the new `spec.scheduling` field.
  2. Enable the `WorkloadWithJob` feature gate and upgrade kube-controller-manager.
  3. New or reconciled Jobs get a `Workload`/`PodGroup` compiled from their `spec.scheduling`,
     defaulting to `Basic` when the field is omitted.

- **Downgrade:**
  1. Disable the feature gate and downgrade kube-controller-manager.

>   2. The API server clears `spec.scheduling` on writes (preserving an already-set value on the stored

Reviewer 3: Only if FG was also disabled on kube-apiserver too (which isn't clear from the point 1).

Author: updated. PTAL.

    their next reconciliation if they do not yet have running pods.
  - To request gang or topology for an existing Job, set `spec.scheduling`. Note that all
    `spec.scheduling` fields except `gang.minCount` are immutable, so changing them on an existing
    Job requires recreating it.

- **Controller restarts and upgrades:**
  - The Job controller only creates `Workload`/`PodGroup` when the Job has no pods.
  - If the controller restarts or is upgraded after creating the `Workload` but before creating the
    `PodGroup` or pods, on the next sync it discovers the existing objects via informers/listers and
    continues without creating duplicates.
  - No special handling is required for in-flight Jobs during controller upgrade or restart.

### Version Skew Strategy

Workload-aware scheduling for a Job only takes effect when:
- The API server can serve the `Workload`/`PodGroup` APIs and accept `spec.scheduling`.
- The Job controller compiles `Workload`/`PodGroup` and sets `schedulingGroup` on pods.
- The scheduler supports `Workload`/`PodGroup`.

Therefore, for a safe rollout:
- kube-apiserver must be upgraded first so it can serve `scheduling.k8s.io/v1alpha3` and persist the

>   new `spec.scheduling` field.

Reviewer 3: spec.scheduling field is not relevant to scheduling.k8s.io/v1alpha3, but rather to the Job API changes in 1.37

- kube-controller-manager can be upgraded before or after the scheduler.

> - kube-scheduler should be upgraded before or together with kube-controller-manager; gang/topology

Reviewer 3: this is conflicting with the sentence above :)

###### Does enabling the feature change any default behavior?
Yes. When the feature gate is enabled:

> 1. The `batch/v1` `spec.scheduling` field becomes available; the API server persists it and defaults

Reviewer 3: This is not a change in default behavior - you need to set the fields appropriate. Please remove this point.

   it to the `Basic` policy when omitted.
2. The Job controller compiles `spec.scheduling` into a `Workload` and `PodGroup` for every root Job
   (one that owns its `Workload`) before creating pods — including `Basic` Jobs.

> 3. Jobs that set `spec.scheduling.policy.gang` use gang scheduling, meaning all `minCount` pods must

Reviewer 3: This is not a default behavior - requires opt-in. Please remove.

>    be scheduled together or none are scheduled.

Reviewer 3: Point (4) is also not relevant now - it requires explicit opt-in.

In summary - the change in default behavior is that the Workload/PodGroup objects are now created and managed for each Job object.

4. The binding of pods referencing a `PodGroup` is delayed until the `PodGroup` object exists.

Note this differs from the v1.36 alpha, where indexed fully-parallel Jobs were gang-scheduled
automatically. In v1.37 gang scheduling is an explicit opt-in via `spec.scheduling`; absent the
field, Jobs default to `Basic`.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes. The feature can be disabled (`WorkloadWithJob: false`). The API server then clears

> `spec.scheduling` on writes (preserving an already-stored value), and the controller stops compiling

Reviewer 3: s/on writes/on creations/

`Workload` and `PodGroup`.

###### What happens if we reenable the feature if it was previously rolled back?

When the feature is re-enabled:
- New and reconciled Jobs without running pods get `Workload` and `PodGroup` compiled from their
  `spec.scheduling` (defaulting to `Basic`) on their next reconciliation cycle.
- Jobs that already have running pods are not affected, since scheduling policy only applies to

>   pods evaluated in future scheduling cycles.

Reviewer 3: What about Jobs for which we created Workload/PodGroup (or e.g. only Workload) before the feature was disabled?

Author: Good point. Updated.

###### How can a rollout or rollback fail? Can it impact already running workloads?

- If the API server doesn't serve the `Workload` and `PodGroup` APIs, the Job controller fails 
  to compile the `Workload` and requeues the Job until the API server is upgraded.
- If the scheduler does not support `Workload` and `PodGroup`, pods are scheduled pod-by-pod 
  with no gang/topology benefit.
- Already running Jobs are not affected by enabling the feature; pods already scheduled and 
  running continue to run. New Jobs or Jobs being reconciled are affected.
- On rollback, disabling the feature gate lets Jobs run without `Workload` or `PodGroup` compilation. 

>   Existing objects are harmless; they are ignored while the feature is disabled.

Reviewer 3: That is not fully true - if you requested a gang, disabling WorkloadOnJob FG doesn't disabled gang-scheduling in the scheduler, so we will probably continue to gang-schedule them. Same for all other features.

The only question is if newly created pods will have a reference to PodGroup...

Author: Updated. PTAL.

### Dependencies

###### Does this feature depend on any specific services running in the cluster?

> - `scheduling.k8s.io/v1alpha3` for the `Workload` and `PodGroup` APIs ([KEP-6089]).

Reviewer 3: L1129 - let's remove it - it's not required for Alpha and it will include also other Workloads/PodGroups.

> - The shared `workloadbuilder` library that compiles user intent into `Workload`/`PodGroup` objects.

Reviewer 3: This is a compile-time dependency - I would remove it.
